### ISSUE TITLE
java.lang.IllegalArgumentException: JSON does not contain Operation Id or Status

### ISSUE THREAD

**Issue by vikramtamboli** [Azure/azure-iot-sdk-java#289]
when we try to registed a device using DPS, it throws below exception.

Waiting for Provisioning Service to register
java.util.concurrent.ExecutionException: java.lang.IllegalArgumentException: JSON does not contain Operation Id or Status
	at java.util.concurrent.FutureTask.report(FutureTask.java:122)
	at java.util.concurrent.FutureTask.get(FutureTask.java:206)
	at com.microsoft.azure.sdk.iot.provisioning.device.internal.task.ProvisioningTask.invokeRegister(ProvisioningTask.java:111)
	at com.microsoft.azure.sdk.iot.provisioning.device.internal.task.ProvisioningTask.call(ProvisioningTask.java:283)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.IllegalArgumentException: JSON does not contain Operation Id or Status
	at com.microsoft.azure.sdk.iot.provisioning.device.internal.parser.RegistrationOperationStatusParser.createFromJson(RegistrationOperationStatusParser.java:71)
	at com.microsoft.azure.sdk.iot.provisioning.device.internal.task.RegisterTask.authenticateWithX509(RegisterTask.java:136)
	at com.microsoft.azure.sdk.iot.provisioning.device.internal.task.RegisterTask.authenticateWithDPS(RegisterTask.java:283)
	at com.microsoft.azure.sdk.iot.provisioning.device.internal.task.RegisterTask.call(RegisterTask.java:308)
	at com.microsoft.azure.sdk.iot.provisioning.device.internal.task.RegisterTask.call(RegisterTask.java:33)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	... 4 more
Registration error, bailing out

**Source Code**

ProvisioningDeviceClient provisioningDeviceClient = null;
            ProvisioningStatus provisioningStatus = new ProvisioningStatus();

            Collection<String> signerCertificates = new LinkedList<>();
            SecurityProvider securityProviderX509 = new SecurityProviderX509Cert(leafPublicPem, leafPrivateKey, signerCertificates);

            provisioningDeviceClient = ProvisioningDeviceClient.create(getGlobalEndpoint(), getScopeId(), getProtocol(), securityProviderX509);

            provisioningDeviceClient.registerDevice(new ProvisioningDeviceClientRegistrationCallbackImpl(), provisioningStatus);

            while (provisioningStatus.provisioningDeviceClientRegistrationInfoClient.getProvisioningDeviceClientStatus() != ProvisioningDeviceClientStatus.PROVISIONING_DEVICE_STATUS_ASSIGNED) {
                if (provisioningStatus.provisioningDeviceClientRegistrationInfoClient.getProvisioningDeviceClientStatus() == ProvisioningDeviceClientStatus.PROVISIONING_DEVICE_STATUS_ERROR
                        || provisioningStatus.provisioningDeviceClientRegistrationInfoClient.getProvisioningDeviceClientStatus() == ProvisioningDeviceClientStatus.PROVISIONING_DEVICE_STATUS_DISABLED
                        || provisioningStatus.provisioningDeviceClientRegistrationInfoClient.getProvisioningDeviceClientStatus() == ProvisioningDeviceClientStatus.PROVISIONING_DEVICE_STATUS_FAILED) {
                    provisioningStatus.exception.printStackTrace();
                    System.out.println("Registration error, bailing out");
                    break;
                }
                System.out.println("Waiting for Provisioning Service to register");
                Thread.sleep(MAX_TIME_TO_WAIT_FOR_REGISTRATION);
            } 

can you please help us to understand what we are missing here?

**Comment by timtay-microsoft** [comment]
This looks like the Device Provisioning Service is sending some json with an unexpected format to this device SDK. You likely aren't doing anything wrong, and we will dig into why the service is behaving this way.

What version of the Provisioning Device Client are you using? Can you repro this bug with the latest version?

**Comment by vikramtamboli** [comment]
we were using version "provisioning-device-client-1.3.0.jar"
now upgraded to "provisioning-device-client-1.3.1.jar"

**Still the error persist**

Waiting for Provisioning Service to register
java.util.concurrent.ExecutionException: java.lang.IllegalArgumentException: JSON does not contain Operation Id or Status
	at java.util.concurrent.FutureTask.report(FutureTask.java:122)
	at java.util.concurrent.FutureTask.get(FutureTask.java:206)
	at com.microsoft.azure.sdk.iot.provisioning.device.internal.task.ProvisioningTask.invokeRegister(ProvisioningTask.java:112)
	at com.microsoft.azure.sdk.iot.provisioning.device.internal.task.ProvisioningTask.call(ProvisioningTask.java:286)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.IllegalArgumentException: JSON does not contain Operation Id or Status
	at com.microsoft.azure.sdk.iot.provisioning.device.internal.parser.RegistrationOperationStatusParser.createFromJson(RegistrationOperationStatusParser.java:71)
	at com.microsoft.azure.sdk.iot.provisioning.device.internal.task.RegisterTask.authenticateWithX509(RegisterTask.java:137)
	at com.microsoft.azure.sdk.iot.provisioning.device.internal.task.RegisterTask.authenticateWithDPS(RegisterTask.java:284)
	at com.microsoft.azure.sdk.iot.provisioning.device.internal.task.RegisterTask.call(RegisterTask.java:310)
	at com.microsoft.azure.sdk.iot.provisioning.device.internal.task.RegisterTask.call(RegisterTask.java:33)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	... 4 more
Registration error, bailing out

**Comment by vikramtamboli** [comment]
Hi JMayrbaeurl

we checked with 
MQTT gave exception "JSON does not contain Operation Id or Status"
AMQPS gave exception "Did not receive DPS registration successfully"

will check with HTTPS as well and let you know if it works.

**Comment by alextolp** [comment]
Could you capture the JSON that is returned? It's likely that the service is returning an error. You could get more info if you enable Diagnostics using the Azure Portal.

**Comment by JMayrbaeurl** [comment]
I still can't reproduce the error. I've now tested with HTTPS, AMQPS and MQTTS. It always worked as expected. Or do you use the Websocket versions of the protocols? Is this an individual or a group enrollment? Getting more logging and diagnostic information would be very helpful. Especially the returned JSON at com.microsoft.azure.sdk.iot.provisioning.device.internal.parser.RegistrationOperationStatusParser.createFromJson(String)

**Comment by vikramtamboli** [comment]
after using HTTPS exact error surfaced out as "Invalid RegistrationId"
certificate had a capital character
seems MQTT & AMQP does not display correct error

we had some environment issues so could not test today
will configure registration id in small characters and test and update

**Comment by timtay-microsoft** [comment]
We'll look into why this error wasn't surfaced by the other protocols. Is it safe to say that you are unblocked for now, though?

**Comment by vikramtamboli** [comment]
Thanks for your help & support.
Yesterday the environment was up and running and successfully provisioned new device using DPS.

Thanks a million!!!

**Comment by timtay-microsoft** [comment]
I'm actually going to leave this issue as open since this sdk was not properly exposing the exception here, just so that we can track this issue and get a proper fix up for it. I'm glad to hear that you are unblocked, though!

**Comment by timtay-microsoft** [comment]
I'm going to close this issue, and re-open a separate issue with the complete bug details, just for tracking purposes.

**Comment by az-iot-builder-01** [comment]
@vikramtamboli, @JMayrbaeurl, @alextolp, thank you for your contribution to our open-sourced project! Please help us improve by filling out this 2-minute [customer satisfaction survey](https://forms.office.com/Pages/ResponsePage.aspx?id=v4j5cvGGr0GRqy180BHbRx8jpG79OudPqY-BojcKjkJUMFFBQjgxNkw5ODJIRjBWTzI0Nk85R0pUQiQlQCN0PWcu)
